contributors/design-proposals/storage/local-volume-management.md

```diff
@@ -0,0 +1,219 @@
+4. Kubelet periodically checks volumes in `DesiredStateOfWorld` to make sure if they are mounted. If not, kubelet will:
+	- Wait for the attach operation finished
+	- Mount the attached device to the global mount path
+	- Bind the global mount path to a pod specified path
+
+This workflow if nicely for all cloud volumes(`AWS EBS`, `GCE PD`, etc.) which have a remote attach API, but not very suitable for local volumes(`fc`, `iscsi`, `rbd`, etc.) which need a local attach operation. So current implementation make the `Attach/Detach` method of these local volumes do nothing, and execute the real attach operations in `WaitForAttach` method. This led to the following problems:
+
+1. Device path missed in `Node.Status.VolumesAttached`.  A/D controller gets volumes' device path from the returned parameter of `Attach` method, and  updates according Node API object. Local volumes' `Attach` method does nothing, so the returned device path is empty. These absent fields may make user confused.
+
+2. Unexpected umount behavior in kubelet. Kubelet will umount the device from node if no pods use a volume, the `UnmountDeviceFunc` will use the device path stored in `Node.Status.VolumesAttached` to check if this device is still opened or not. As this device path is empty for local volumes, there may be unexpected behaviors.
+
+3. Complex and confusing code both in kubernetes side(`kubelet Volumemanager` and `A/D controller`) and volume plugin side, as local volumes have to implement `AttachableVolumePlugin` interface.
+
+## <a name="object"></a>Object
+
+To solve the problems as above, we need to split the `AttachableVolumePlugin` interface into two independent interfaces: One is still called `AttachableVolumePlugin` and another is `DeviceMountableVolumePlugin` . The new `AttachableVolumePlugin` has the same interface with the old one and implemented by the cloud volumes only, `DeviceMountableVolumePlugin` has the device mount/umount interface only and implemented by the local volumes. Accordingly, we also need to split `Attacher` intreface to `Attacher` and `DeviceMounter`, `Detacher` to `Detacher` and `DeviceUmounter`.
+
+## <a name="roadmap"></a>Roadmap
+
+* Volume relative interface refactor
+	* Refactor volume plugin interfaces(`AttachableVolumePlugin`) and volume interfaces(`Attacher/Detacher`)
+	* Refactor `A/D controller` and `Kubelet VolumeManager` to use new interfaces
+* Refactor local volumes to implement `DevicemountableVolumePlugin` only
+	* Refactor `rbd`
+	* Refactor `fc`, `iscsi`, etc.
+
+## <a name="design"></a>Implementation Design
+
+### <a name="k8s_refactor"></a>Kubernetes Side Refactor
+
+#### <a name="k8s_interfaces_refactor"></a>Volume Plugin Interfaces refactor
+
+We extract the device mount relative methods from `Attacher` to a new interface, and embed this new interface back to `Attacher` so that volumes implement `Attacher` need also implement `DeviceMounter`:
```

Comment:
"volumes implement `Attacher` need also implement `DeviceMounter`" - that's a good thing to require, however I don't think that embedding `DeviceMounter` into `Attacher` is the right thing to do. I'd leave the interface independent and check it in runtime (i.e. in `GenerateMountVolumeFunc`).